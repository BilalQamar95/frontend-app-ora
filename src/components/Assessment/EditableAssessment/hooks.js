import { useContext } from 'react';
import { StrictDict } from '@edx/react-unit-test-utils';

import { useRubricConfig } from 'data/services/lms/hooks/selectors';
import { useSubmitRubric } from 'data/services/lms/hooks/actions';
import { AssessmentContext } from 'components/AssessmentContext';

export const stateKeys = StrictDict({
  optionsSelected: 'optionsSelected',
  criterionFeedback: 'criterionFeedback',
  assessment: 'assessment',
  overallFeedback: 'overallFeedback',
});

const useEditableAssessmentData = () => {
  const {
    currentValue,
    formFields,
  } = useContext(AssessmentContext);

  const { criteria, feedbackConfig } = useRubricConfig();

  const submitRubricMutation = useSubmitRubric();
  const onSubmit = () => {
    submitRubricMutation.mutate(currentValue);
  };

  return {
    criteria,
    formFields,
    onSubmit,
    submitStatus: submitRubricMutation.status,
    // overall feedback
    overallFeedbackPrompt: feedbackConfig.defaultText,
  };
};

export default useEditableAssessmentData;
