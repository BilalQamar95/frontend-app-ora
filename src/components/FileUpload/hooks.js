import { useCallback } from 'react';

import { StrictDict, useKeyedState } from '@edx/react-unit-test-utils';

export const stateKeys = StrictDict({
  errors: 'errors',
  isModalOpen: 'isModalOpen',
  uploadArgs: 'uploadArgs',
  description: 'description',
});
export const useUploadConfirmModalHooks = ({
  files, closeHandler, uploadHandler,
}) => {
  const [description, setDescription] = useKeyedState(stateKeys.description, null);
  const [errors, setErrors] = useKeyedState(stateKeys.errors, []);

  const confirmUploadClickHandler = () => {
    const errorList = files.map((file) => (!file.description));
    setErrors(errorList);
    if (errorList.some((error) => error)) {
      return;
    }
    uploadHandler();
  };

  const exitHandler = () => {
    setErrors([]);
    closeHandler();
  };

  // Modifying pointer of file object. This is not a good practice.
  // eslint-disable-next-line no-param-reassign, no-return-assign
  const onFileDescriptionChange = (event) => setDescription(event.target.value);

  return {
    errors,
    confirmUploadClickHandler,
    exitHandler,
    onFileDescriptionChange,
  };
};

export const useFileUploadHooks = ({
  onFileUploaded,
}) => {
  const [uploadArgs, setUploadArgs] = useKeyedState(stateKeys.uploadArgs, {});
  const [isModalOpen, setIsModalOpen] = useKeyedState(stateKeys.isModalOpen, false);

  const confirmUpload = useCallback(async () => {
    setIsModalOpen(false);
    if (onFileUploaded) {
      await onFileUploaded(uploadArgs);
    }
    setUploadArgs({});
  }, [uploadArgs, onFileUploaded, setIsModalOpen, setUploadArgs]);

  const closeUploadModal = useCallback(() => {
    setIsModalOpen(false);
    setUploadArgs({});
  }, [setIsModalOpen, setUploadArgs]);

  const onProcessUpload = useCallback(({ fileData, handleError, requestConfig }) => {
    setIsModalOpen(true);
    setUploadArgs({ fileData, handleError, requestConfig });
  }, [setIsModalOpen, setUploadArgs]);

  return {
    isModalOpen,
    uploadArgs,
    confirmUpload,
    closeUploadModal,
    onProcessUpload,
  };
};

export default {
  useUploadConfirmModalHooks,
  useFileUploadHooks,
};
