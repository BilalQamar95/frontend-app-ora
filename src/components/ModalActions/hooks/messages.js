import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  finishLater: {
    id: 'ora-mfe.ModalActions.simpleAction.finishLater',
    defaultMessage: 'Save for later',
    description: 'Save for later (close) button text',
  },
  responseSubmitted: {
    id: 'ora-mfe.ModalActions.responseSubmitted',
    defaultMessage: 'Response submitted',
    description: 'Submit button text after successful submission',
  },
  savingResponse: {
    id: 'ora-mfe.ModalActions.savingResponse',
    defaultMessage: 'Saving response',
    description: 'Save for later button text while saving',
  },
});

export default messages;
