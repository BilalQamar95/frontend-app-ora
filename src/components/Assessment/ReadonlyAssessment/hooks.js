import { useRubricConfig } from 'data/services/lms/hooks/selectors';

export const useReadonlyAssessmentData = ({ assessment }) => {
  const { criteria, feedbackConfig } = useRubricConfig();

  const criterionData = (criterion) => ({
    ...criterion,
    optionsValue: assessment.optionsSelected[criterion.name],
    optionsIsInvalid: false,
    feedbackValue: assessment.criterionFeedback[criterion.name],
    feedbackIsInvalid: false,
  });

  return {
    criteria: criteria.map(criterionData),
    // overall feedback
    overallFeedbackPrompt: feedbackConfig.defaultText,
    overallFeedback: assessment.overallFeedback,
  };
};

export default useReadonlyAssessmentData;
