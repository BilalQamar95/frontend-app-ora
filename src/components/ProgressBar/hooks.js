import { useParams } from 'react-router-dom';
import { useIsEmbedded, useViewStep } from 'hooks';
import {
  useEffectiveGrade,
  useGlobalState,
} from 'data/services/lms/hooks/selectors';
import {
  stepRoutes,
  stepStates,
} from 'data/services/lms/constants';

export const useProgressStepData = ({ step, canRevisit = false }) => {
  const { xblockId, courseId } = useParams();
  const isEmbedded = useIsEmbedded();
  const viewStep = useViewStep();
  const { stepState } = useGlobalState({ step });

  const href = `/${stepRoutes[step]}${isEmbedded ? '/embedded' : ''}/${courseId}/${xblockId}`;
  const isActive = viewStep === step;
  const isEnabled = (
    isActive
    || (stepState === stepStates.inProgress)
    || (canRevisit && stepState === stepStates.completed)
  );
  const myGrade = useEffectiveGrade()?.stepScore;
  return {
    href,
    isEnabled,
    isActive,
    isComplete: stepState === stepStates.completed,
    inProgress: stepState === stepStates.inProgress,
    isPastDue: stepState === stepStates.closed,
    myGrade,
    // myGrade: { earned: 8, possible: 10 },
    // isPastDue: step === 'self',
  };
};

export default { useProgressStepData };
