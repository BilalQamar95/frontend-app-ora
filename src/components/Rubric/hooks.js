import { StrictDict } from 'utils';
import { useRubricConfig } from 'data/services/lms/hooks/selectors';

export const ButtonStates = StrictDict({
  default: 'default',
  pending: 'pending',
  complete: 'complete',
  error: 'error',
});

export const useRubricData = ({
  isGrading,
}) => {
  const criteria = useRubricConfig().criteria.map(
    (_, index) => ({
      isGrading,
      key: index,
      orderNum: index,
    }),
  );
  const submitButtonState = ButtonStates.default;

  return {
    criteria,
    showFooter: isGrading,
    buttonProps: {
      onClick: () => ({}),
      state: submitButtonState,
      disabledStates: [ButtonStates.pending, ButtonStates.complete],
    },
  };
};
