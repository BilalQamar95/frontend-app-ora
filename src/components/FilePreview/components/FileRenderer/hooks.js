import { useKeyedState, StrictDict } from '@edx/react-unit-test-utils';

import { errorStatuses, errorMessages, renderers } from '../constants';
import { getFileType } from '../utils';
import messages from './messages';

export const stateKeys = StrictDict({
  errorStatus: 'errorStatus',
  isLoading: 'isLoading',
});

/**
 * component hooks
 */
export const useRenderData = ({
  file,
  formatMessage,
}) => {
  const [errorStatus, setErrorStatus] = useKeyedState(stateKeys.errorStatus, null);
  const [isLoading, setIsLoading] = useKeyedState(stateKeys.isLoading, true);

  const setState = (newState) => {
    setErrorStatus(newState.errorStatus);
    setIsLoading(newState.isLoading);
  };

  const stopLoading = (status = null) => setState({ isLoading: false, errorStatus: status });

  const errorMessage = (
    errorMessages[errorStatus] || errorMessages[errorStatuses.serverError]
  );
  const errorAction = {
    id: 'retry',
    onClick: () => setState({ errorStatus: null, isLoading: true }),
    message: messages.retryButton,
  };
  const error = {
    headerMessage: errorMessage,
    children: formatMessage(errorMessage),
    actions: [errorAction],
  };

  const Renderer = renderers[getFileType(file.fileName)];
  const rendererProps = {
    fileName: file.fileName,
    url: file.fileUrl,
    onError: stopLoading,
    onSuccess: () => stopLoading(),
  };

  return {
    errorStatus,
    isLoading,
    error,
    Renderer,
    rendererProps,
  };
};
