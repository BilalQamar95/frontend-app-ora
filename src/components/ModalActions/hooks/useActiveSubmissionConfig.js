import React from 'react';

import { MutationStatus } from 'constants';
import messages from './messages';

const useActiveSubmissionConfig = ({
  options = {},
  closeModal,
  formatMessage,
}) => {
  const saveAndClose = React.useCallback(
    () => (options.finishLater ? options.finishLater().then(closeModal) : null),
    [options, closeModal],
  );

  if (!options) { return null; }

  const { submit, submitStatus, finishLaterStatus } = options;

  return {
    primary: {
      onClick: submit,
      state: submitStatus,
      labels: {
        [MutationStatus.idle]: formatMessage(messages.submitResponse),
        [MutationStatus.loading]: formatMessage(messages.submittingResponse),
      },
    },
    secondary: {
      onClick: saveAndClose,
      state: finishLaterStatus,
      labels: {
        [MutationStatus.idle]: formatMessage(messages.finishLater),
        [MutationStatus.loading]: formatMessage(messages.savingResponse),
      },
    },
  };
};
export default useActiveSubmissionConfig;
