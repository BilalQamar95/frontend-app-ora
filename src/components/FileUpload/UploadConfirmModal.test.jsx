import { shallow } from '@edx/react-unit-test-utils';
import UploadConfirmModal from './UploadConfirmModal';

import { useUploadConfirmModalHooks } from './hooks';

jest.mock('./hooks', () => ({
  useUploadConfirmModalHooks: jest.fn(),
}));

describe('<UploadConfirmModal />', () => {
  const props = {
    open: true,
    file: { name: 'file1' },
    closeHandler: jest.fn().mockName('closeHandler'),
    uploadHandler: jest.fn().mockName('uploadHandler'),
  };

  const mockHooks = (overrides) => {
    useUploadConfirmModalHooks.mockReturnValueOnce({
      shouldShowError: false,
      exitHandler: jest.fn().mockName('exitHandler'),
      confirmUploadClickHandler: jest.fn().mockName('confirmUploadClickHandler'),
      onFileDescriptionChange: jest.fn().mockName('onFileDescriptionChange'),
      ...overrides,
    });
  };
  describe('renders', () => {
    test('without error', () => {
      mockHooks(
        { errors: new Array(2) },
      );
      const wrapper = shallow(<UploadConfirmModal {...props} />);
      expect(wrapper.snapshot).toMatchSnapshot();

      expect(wrapper.instance.findByType('Form.Group').length).toBe(1);
      expect(wrapper.instance.findByType('Form.Control.Feedback').length).toBe(0);
    });

    test('with errors', () => {
      mockHooks({ shouldShowError: true });
      const wrapper = shallow(<UploadConfirmModal {...props} />);
      expect(wrapper.snapshot).toMatchSnapshot();

      expect(wrapper.instance.findByType('Form.Group').length).toBe(1);
      expect(wrapper.instance.findByType('Form.Control.Feedback').length).toBe(1);
    });
  });
});
