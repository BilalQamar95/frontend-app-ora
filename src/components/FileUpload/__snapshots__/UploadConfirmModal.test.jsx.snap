// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UploadConfirmModal /> renders with errors 1`] = `
<ModalDialog
  hasCloseButton={false}
  isBlocking={true}
  isOpen={true}
  onClose={[MockFunction exitHandler]}
  title="Add a text description to your file"
>
  <ModalDialog.Header>
    <ModalDialog.Title>
      Add a text description to your file
    </ModalDialog.Title>
  </ModalDialog.Header>
  <ModalDialog.Body>
    <div>
      <Form.Group>
        <FormLabel>
          <strong>
            Description for: 
          </strong>
          <span
            className="file-name-ellipsis"
          >
            file1
          </span>
        </FormLabel>
        <Form.Control
          isInvalid={true}
          name="file-description"
          onChange={[MockFunction onFileDescriptionChange]}
        />
        <Form.Control.Feedback
          type="invalid"
        >
          formatMessage(messages.fileDescriptionMissingError)
        </Form.Control.Feedback>
      </Form.Group>
    </div>
  </ModalDialog.Body>
  <ModalDialog.Footer>
    <ActionRow>
      <ModalDialog.CloseButton
        onClick={[MockFunction exitHandler]}
        variant="tertiary"
      >
        Cancel upload
      </ModalDialog.CloseButton>
      <Button
        onClick={[MockFunction confirmUploadClickHandler]}
        variant="primary"
      >
        Upload files
      </Button>
    </ActionRow>
  </ModalDialog.Footer>
</ModalDialog>
`;

exports[`<UploadConfirmModal /> renders without error 1`] = `
<ModalDialog
  hasCloseButton={false}
  isBlocking={true}
  isOpen={true}
  onClose={[MockFunction exitHandler]}
  title="Add a text description to your file"
>
  <ModalDialog.Header>
    <ModalDialog.Title>
      Add a text description to your file
    </ModalDialog.Title>
  </ModalDialog.Header>
  <ModalDialog.Body>
    <div>
      <Form.Group>
        <FormLabel>
          <strong>
            Description for: 
          </strong>
          <span
            className="file-name-ellipsis"
          >
            file1
          </span>
        </FormLabel>
        <Form.Control
          isInvalid={false}
          name="file-description"
          onChange={[MockFunction onFileDescriptionChange]}
        />
      </Form.Group>
    </div>
  </ModalDialog.Body>
  <ModalDialog.Footer>
    <ActionRow>
      <ModalDialog.CloseButton
        onClick={[MockFunction exitHandler]}
        variant="tertiary"
      >
        Cancel upload
      </ModalDialog.CloseButton>
      <Button
        onClick={[MockFunction confirmUploadClickHandler]}
        variant="primary"
      >
        Upload files
      </Button>
    </ActionRow>
  </ModalDialog.Footer>
</ModalDialog>
`;
